mkswap: implement --file

endrift opened the pull request:
Addresses #2359

I don't really think this is a sufficient number of tests, but it seems like there already is a pretty lackluster number of tests for mkswap.

masatake commented on disk-utils/mkswap.c:
```diff
@@ -351,13 +356,17 @@ static unsigned long long get_size(const struct mkswap_control *ctl)
-	if ((unsigned long long) ctl->offset > size)
-		errx(EXIT_FAILURE, _("offset larger than file size"));
+	if (ctl->file && ctl->size)
+		size = ctl->size;
```
If the control enters this branch, fd is not initialized.
So close(fd) at line 364/373 causes an unexpected behavior.
A condition guarding the invocation of close needs. e.g.:
```
if (! (ctl->file && ctl->size))
   close(fd);
```

karelzak replied:
I think `fd` can be defined and used only in the `else { }` block, the close(fd) could be at the end of the block.

karelzak commented on disk-utils/mkswap.8.adoc:
```diff
@@ -76,6 +79,9 @@ Specify the _ENDIANNESS_ to use, valid arguments are *native*, *little* or *big*
 
+*-s*, *--size* _size_::
+Specify the size of the created swap file, in bytes. This option only makes sense when used with *--file*.
+
```
It would be probably better to change the current SYNOPSIS (and related man page information) from
```
 mkswap [options] device [size]
```
to
```
 mkswap [options] device [blocks]
```
It's confusing if we use `size` for two different things.

karelzak replied:
Maybe you can also add complete SYNOPSIS line for the file
```
  mkswap [options] device [blocks]
  mkswap [options] --file filename --size size
```

karelzak commented on disk-utils/mkswap.c:
```diff
@@ -88,14 +88,17 @@ struct mkswap_control {
 	char			*opt_label;	/* LABEL as specified on command line */
 	unsigned char		*uuid;		/* UUID parsed by libbuuid */
 
+	unsigned long long	size;		/* desired swap file size */
```
Maybe make it more obvious and use the name `filesz` (or so).

karelzak commented on disk-utils/mkswap.c:
```diff
@@ -382,11 +391,28 @@ static void open_device(struct mkswap_control *ctl)
+			if (chmod(ctl->devname, 0600) < 9)
+				err(EXIT_FAILURE, _("cannot set permissions on swap file %s"), ctl->devname);
+		}
+		ctl->fd = open(ctl->devname, O_RDWR | O_CREAT, 0600);
```
Please add to the man page info that "If the file already exists, it will be truncated to the specified size".

karelzak commented on disk-utils/mkswap.c:
```diff
@@ -382,11 +391,28 @@ static void open_device(struct mkswap_control *ctl)
 		err(EXIT_FAILURE, _("cannot open %s"), ctl->devname);
+	if (ctl->file && ctl->size) {
+		if (ftruncate(ctl->fd, ctl->size) < 0)
+			err(EXIT_FAILURE, _("couldn't allocate swap file %s"), ctl->devname);
```
#ifdef HAVE_POSIX_FALLOCATE
...
#endif

karelzak commented on disk-utils/mkswap.c:
```diff
@@ -588,6 +616,13 @@ int main(int argc, char **argv)
 					10, _("Invalid offset"), SINT_MAX(off_t));
 			break;
+		case 's':
+			ctl.size = str2unum_or_err(optarg,
```
Please, strtosize_or_err() to make it possible to use `--size 100MiB` (and don't forget to add a note about possible suffixes to the man page (see another pages), we usually use
```  
       The size arguments may be followed by the multiplicative suffixes KiB (=1024), MiB
       (=1024*1024), and so on for GiB, TiB, PiB, EiB, ZiB, and YiB (the "iB" is optional, e.g., "K" has the
       same meaning as "KiB") 
```

endrift commented:
Thanks for the review. I've updated this PR per the suggestions, as well as adding a Linux/glibc fallocate path. I'm unsure when posix_fallocate wouldn't be defined, but I added the #ifdef regardless.

endrift commented:
I'm not clear what's leading to this build failure, or if it's even caused by this PR.

karelzak commented:
It seems the coverall test problem is unrelated to your work, let's ignore it for now.
